### ISSUE TITLE
How to set environment variable when package in installed

### ISSUE THREAD

**Issue by jngd** [conan-io/conan#2037]
I'm trying to set env variable that is needed at runtime for what library that I have conified. To automate it, I want that the package write one runtime variable in system env when it is installed, with the route of the package. I have tried with this without luck:

```
def package_info(self):
    self.cpp_info.libs = tools.collect_libs(self)
    self.env_info.MY_VAR = "VALUE"                                     

```
Any idea how to achieve it? I need that this variable persist, in order to use library with this env var when I run my application.

**Comment by memsharded** [comment]
Conan will never modify your system. There is a way to use environment variables without doing that: using environments. You can generate them, declaring them in the conanfile, or even in the command line:

```bash
$ conan install . -g virtualenv
```
This will generate some activate/deactivate scripts for you. You can use them to actually enable the environment variables:

```bash
$ activate.bat # or source activate.sh
//MY_VAR will be defined now
```
Environments can be composed. I specially like the ``virtualrunenv`` that sets the execution paths for apps, even without declaring anything in the package recipe.

Please have a look: http://docs.conan.io/en/latest/reference/generators/virtualenv.html and tell me if this helps. Thanks!

**Comment by jngd** [comment]
Many thanks @memsharded . It helps, but I want to automate it so that when other dev download this package, MY_VAR is set in his/her system, with the route of the package just downloaded, not just run one activate script manually. Can I get package path in activate.sh script?

**Comment by memsharded** [comment]
If you mean to set-up permanently such env-var in the system, it is basically not possible from a python subprocess, not because it is conan. You can have a look at StackOverflow, and you will get some pointer how it is possible to do it in windows calling ``SETX`` or directly writing in the registry. Similary, writing in ``.bashrc`` in linux, to append the environment variable. But overall quite error-prone.

Moreover, this behavior is very likely to be problematic and mess with the system, so doesn't make sense that conan automated it. For sure you can add a bunch of python code, but not to the ``package_info()`` method, if any, I would suggest, in the ``deploy()`` method. This deploy() method is called when you execute:

```bash
$ conan install <pkg-reference> # Like FullPkgName/version@user/channel
```
That code would try to set the environment variable.

Otherwise, yes, you can just set:

```python
def package_info(self):
     self.env_info.MY_PKG_PATH = self.package_folder
```
And you will get a MY_PKG_PATH env-var with the package path in the ``activate.sh`` script.

A different approach, of course, would be to use ``deploy()`` to do a real copy of the binary artifacts to the system.

**Comment by jngd** [comment]
@memsharded you're right, this is a very problematic behavior. I'm thinking that I can copy all libs that are in lib directory (and subdirectories) to the directory that I can read from os env, using this code:

```    
dest = os.getenv("MYVAR", "")
if not dest:
    return
self.copy("*so.*", dst=dest, src="lib/subdir", ignore_case=True)
```
Not whole process is automated, but user just need to have MYVAR in his/her system env. This is just for Linux. Which section I should use to do that?

**Comment by jngd** [comment]
I have tried doing it in imports section. How I can get that one package execute imports section for all of it'r requires?

**Comment by memsharded** [comment]
The imports work in the conanfile (either .txt or .py) which is in user folder, and executing ``conan install .`` over it. That is, the imported files are defined by the consumer, not by the packages themselves. If you want some package to copy on user folders when installed with ``conan install <pkg-ref>``, then you should use the ``deploy()`` method.

**Comment by jngd** [comment]
@memsharded But deploy() method is not transitive, so how I can execute deploy() when one library is required from other? I mean, I want to do deploy() of library called "foo/1.0.0" when I do install for library "bar/1.0.0" which have "foo/1.0.0"  as required.

**Comment by memsharded** [comment]
Neither ``imports()``. Imports is not executed transitively, the one imports() in your consumer conanfile will imports from all the dependency graph, irrespective of what other package recipes have in their imports().

``deploy()`` using the ``copy_deps()`` method, will copy from the dependencies too. 

P.S. Are you in CppLang slack? There is a #conan channel there.

**Comment by jngd** [comment]
Ok, deploy() do the trick. Thanks @memsharded for you attention and thank you also for the slack channel info.
